nixos/modules/services/security/bitwarden-directory-connector.nix

```
@@ -0,0 +1,349 @@
+          description = lib.mdDoc "Attribute for a name of group.";
+          example = "cn";
+        };
+        filter = mkOption {
+          type = types.str;
+          description = lib.mdDoc "Filter for groups.";
+          example = "(cn=sales)";
+        };
+      };
+    };
+
+    env = {
+      ldap = mkOption rec {
+        type = types.str;
+        description = "Auth for the LDAP, has value defined in {option}`ldap.pw_env";
+      };
+      bitwarden = mkOption rec {
+        type = types.str;
+        description = "Auth for Bitwarden, has BW_CLIENTID and BW_CLIENTSECRET";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.groups."${cfg.user}" = {};
+
+    users.users."${cfg.user}" = {
+      createHome = true;
+      isSystemUser = true;
+      home = "${cfg.directory}";
+      group = "${cfg.user}";
+      homeMode = "711";
+    };
+
+    systemd = {
+      timers.bitwarden_directory_connector = {
+        description = "Sync timer for Bitwarden Directory Connector";
+        wantedBy = ["timers.target"];
+        partOf = ["bitwarden_directory_connector.service"];
+        timerConfig = {
+          OnCalendar = cfg.sync.interval;
+          Unit = "bitwarden_directory_connector.service";
+          Persistent = true;
+        };
+      };
+
+      services.bitwarden_directory_connector = {
+        description = "Main process for Bitwarden Directory Connector";
+        wantedBy = ["multi-user.target"];
+        after = ["network-online.target"];
+        wants = [];
+        path = [pkgs.jq];
+
+        environment = {
+          BITWARDENCLI_CONNECTOR_APPDATA_DIR = cfg.directory;
+          BITWARDENCLI_CONNECTOR_PLAINTEXT_SECRETS = "true";
+        };
+
+        serviceConfig = {
+          Type = "oneshot";
+          User = "${cfg.user}";
+          Group = "${cfg.user}";
+          ExecStartPre = pkgs.writeShellScript "bitwarden_directory_connector-config" ''
+            # create the config file
+            ${cfg.package}/bin/${cfg.binary_name} data-file
+
+            ${cfg.package}/bin/${cfg.binary_name} config server ${cfg.domain}
+
+            # now login to set credentials
+            ${cfg.package}/bin/${cfg.binary_name} login
+
+            # set the ldap details
+            account=$(jq '.authenticatedAccounts[0]?' ${cfg.directory}/data.json)
+            jq ".[$account].directoryConfigurations.ldap |= ${json_string ldap_data}" ${cfg.directory}/data.json > ${cfg.directory}/data1.json
+
+            # remove the original
+            rm -f ${cfg.directory}/data.json
+
+            # set the client id
+            orgID=$(echo $BW_CLIENTID | sed 's/organization\.//g')
+            jq ".[$account].directorySettings.organizationId |= \"$orgID\" " ${cfg.directory}/data1.json > ${cfg.directory}/data2.json
+
+            # and sync data
+            jq ".[$account].directorySettings.sync |= ${json_string sync_data}" ${cfg.directory}/data2.json > ${cfg.directory}/data.json
```

Rather than pasting data directly into the `jq` script and trying to make sure it's escaped correctly, I'd use `jq`'s `--arg` and `--argjson` options, together with `lib.escapeShellArg`.

```suggestion
            account=$(jq '.authenticatedAccounts[0]?' ${cfg.directory}/data.json)
            jq '.[$account].directoryConfigurations.ldap |= $ldap_data' --arg account "$account" --argjson ldap_data ${lib.escapeShellArg ldap_data} ${cfg.directory}/data.json > ${cfg.directory}/data1.json

            # remove the original
            rm -f ${cfg.directory}/data.json

            # set the client id
            orgID=$(echo $BW_CLIENTID | sed 's/organization\.//g')
            jq '.[$account].directorySettings.organizationId |= $orgID' --arg account "$account" --arg orgID "$orgID" ${cfg.directory}/data1.json > ${cfg.directory}/data2.json

            # and sync data
            jq '.[$account].directorySettings.sync |= $sync_data' --arg account "$account" --argjson sync_data ${lib.escapeShellArg sync_data} ${cfg.directory}/data2.json > ${cfg.directory}/data.json
```

---

Thats what I was trying to do for a good chunk of the day, couldnt for teh life of me get it to work.  
I'll test out what ye suggest and hopefully that works.

---

Its almost wild how many functions like ``escapeShellArg`` are really useful but discoverability is not great.  
Thanks for this.

---

Oh, I did my suggestion a little wrong. I misread and thought the bash variable `$account` contained a raw string. Looks like it contains a json string literal.

Honestly, this whole section should be condensed down to a single `jq` call, probably.

```suggestion
            jq '.authenticatedAccounts[0] as $account
              | .[$account].directoryConfigurations.ldap |= $ldap_data
              | .[$account].directorySettings.organizationId |= $orgID
              | .[$account].directorySettings.sync |= $sync_data' \
              --argjson ldap_data ${lib.escapeShellArg \
              --arg orgID "''${BW_CLIENTID//organization.}" \
              --argjson sync_data ${lib.escapeShellArg sync_data} \
              ${escapeShellArg cfg.directory}/data.json \
              > ${escapeShellArg cfg.directory}/data.json.tmp
            mv -f -- ${escapeShellArg cfg.directory}/data.json.tmp ${escapeShellArg cfg.directory}/data.json
```

---

I just finished turning ldap and sync into submodules, so this is perfect.  
I'll get to testing this now

---

That worked scarily well.  
``escapeShellArg cfg.directory`` is in case the user has their path in quotes right (``"/path to/dir"``).  
``BW_CLIENTID//organization.`` never seen this before, where ye cna call an env var and strip it all in one go, neat.  
Overall seems like ye are quite used to ``jq``, I am taking notes :D

---

Changes pushed for all suggestions

---

> `escapeShellArg cfg.directory` is in case the user has their path in quotes right (`"/path to/dir"`).

It's in case the path has problematic characters like spaces or other shell-meaningful characters in it. Not sure if that's what you meant about quotes.
> `BW_CLIENTID//organization.` never seen this before, where ye cna call an env var and strip it all in one go, neat. 
Yeah, bash has a whole set of modifications you can make to shell or environment variables when you expand them. Check the `Parameter Expansion` section of `man bash'.

---

> Not sure if that's what you meant about quotes.

Ye it is what I meant (I like using spaces in a path as an example as its normally one of the first issues folks run into in my course)